Blog Post on Flink 1.10 Python UDF support (#308)
apache/flink-web: _posts/2020-02-23-pyflink-udf-support-flink.md

---
layout: post
title: "PyFlink: Introducing Python Support for UDFs in Flink's Table API"
date: 2020-02-23T12:00:00.000Z
authors:
- Jincheng:
  name: "Jincheng Sun"

> excerpt: Flink1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink. 

Reviewer: ```suggestion
excerpt: Flink 1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink. 
```

---

> Flink 1.9 introduced the Python Table API, allowing developers and data engineers to write Python Table API jobs for Table Transformations and Analysis, such as Python ETL or Aggregate jobs. However, Python users faced some limitations when it came to support for Python UDFs in Flink 1.9, preventing them from extending the system’s built-in functionality. 

Reviewer: ```suggestion
Flink 1.9 introduced the Python Table API, allowing developers and data engineers to write Python Table API jobs for Table transformations and analysis, such as Python ETL or aggregate jobs. However, Python users faced some limitations when it came to support for Python UDFs in Flink 1.9, preventing them from extending the system’s built-in functionality. 
```

> In Flink1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF Environment and Dependency Management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 

Reviewer 2: ```suggestion
In Flink 1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF Environment and Dependency Management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 
```

> In Flink1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF Environment and Dependency Management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 

Reviewer: ```suggestion
In Flink 1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF environment and dependency management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 
```

  * The Python UDF Runner sends the user defined functions to be executed in the Python worker.
  * The Python worker transforms the user defined functions to Beam operations. (Note that we leverage the power of [Beam’s Portability Framework](https://beam.apache.org/roadmap/portability/) to execute the Python UDF.)
  * The Python worker establishes the gRPC connections, such as the data connection, logging connection, etc.

2. Processing of the input elements.
  * The Python UDF Runner sends the input elements via the gRPC data service to the Python worker for execution.
  * The Python use defined function can access the state via gRPC state service during execution.
  * The Python user defined function can also aggregate the logging and metrics to the Python UDF Runner via the gRPC logging service and the metrics service during execution.
  * The execution results are finally sent to the Python UDF Runner via the gRPC data service.

# How to use PyFlink with UDFs in Flink 1.10

This section provides some Python user defined function (UDF) examples, including how to install PyFlink, how to define/register/invoke UDFs in PyFlink and how to execute the job.

> ### Install PyFlink

Reviewer 3: Move to heading level 2?

> ### Install PyFlink

Author: ```suggestion
## Install PyFlink
```

Using Python in Apache Flink requires installing PyFlink. PyFlink is available through PyPi and can be easily installed using pip: 

> ```python

Reviewer 3: The installation commands are missing here (and maybe something else?)

Reviewer 3: Add a heading for how to use Python UDFs

Reviewer 3: Maybe first show a simple but complete Pyflink program without a UDF?

Reviewer 4: > The installation commands are missing here (and maybe something else?)

+1, I think the command `python -m pip install apache-flink`  should be added  .  same as `Install PyFlink` part in [1].
[1] https://docs.google.com/document/d/1I83o8e4CSu1baKPK9SKq1S1GN-p98AVvDOIgUHQVOZo/edit# 

> Add a heading for how to use Python UDFs

This post is talk about Python UDF support, So I am fine without a simple example. :)

                 .field('b', DataTypes.BIGINT())) \
    .with_schema(Schema()
                 .field('a', DataTypes.BIGINT())
                 .field('b', DataTypes.BIGINT())) \
    .create_temporary_table('mySource')

t_env.connect(FileSystem().path('/tmp/output')) \
    .with_format(OldCsv()
                 .field('sum', DataTypes.BIGINT())) \
    .with_schema(Schema()
                 .field('sum', DataTypes.BIGINT())) \
    .create_temporary_table('mySink')

t_env.from_path('mySource')\
    .select("add(a, b)") \
    .insert_into('mySink')

t_env.execute("tutorial_job")
```

### Submit the job

Firstly, you need to prepare the input data in the “/tmp/input” file. For example,

`$ echo "1,2" > /tmp/input`

Next, you can run this example on the command line,

`$ python python_udf_sum.py`

The command builds and runs the Python Table API program in a local mini-cluster. You can also submit the Python Table API program to a remote cluster using different command lines, (see more details [here](https://ci.apache.org/projects/flink/flink-docs-master/ops/cli.html#job-submission-examples)).

Finally, you can see the execution result on the command line:

`$ cat /tmp/output
 3`

> ### Python UDF dependency management

Reviewer 3: move this to heading 2 for more structure?

> ### Python UDF dependency management

Author: ```suggestion
## Python UDF dependency management
```

from pyflink.table.udf import udf

env = StreamExecutionEnvironment.get_execution_environment()
env.set_parallelism(1)
t_env = StreamTableEnvironment.create(env)

@udf(input_types=[DataTypes.BIGINT(), DataTypes.BIGINT()], result_type=DataTypes.BIGINT())
def add(i, j):
    from mpmath import fadd
    return int(fadd(1, 2))

t_env.set_python_requirements("/tmp/requirements.txt", "/tmp/cached_dir")
t_env.register_function("add", add)

t_env.connect(FileSystem().path('/tmp/input')) \
    .with_format(OldCsv()
                 .field('a', DataTypes.BIGINT())
                 .field('b', DataTypes.BIGINT())) \
    .with_schema(Schema()
                 .field('a', DataTypes.BIGINT())
                 .field('b', DataTypes.BIGINT())) \
    .create_temporary_table('mySource')

t_env.connect(FileSystem().path('/tmp/output')) \
    .with_format(OldCsv()
                 .field('sum', DataTypes.BIGINT())) \
    .with_schema(Schema()
                 .field('sum', DataTypes.BIGINT())) \
    .create_temporary_table('mySink')

t_env.from_path('mySource')\
    .select("add(a, b)") \
    .insert_into('mySink')

t_env.execute("tutorial_job")
```

> ### Submit the job

Reviewer 3: we had the same heading before. This is a bit confusing

Reviewer 4: May be `Submit a job with dependency` ?

> ### Submit the job

Author: ```suggestion
### Submit a job with dependency 
```
